### PR TITLE
Feat/support block quant

### LINKED ISSUE #879: [Feature] Support block-wise quantization for Linear and EPMoE on TPU
### Checklist

- [x] 1. If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sgl-jax/discussions/new/choose Otherwise, it will be closed.
- [x] 2. Please use English, otherwise it will be closed.

### Motivation

Current quantization support in sglang-jax only covers per-channel (per-tensor) quantization. Many models (e.g., DeepSeek-V3, Qwen3, MiMo-v2-flash) ship with block-wise quantized checkpoints using 128x128 block scales. Without block-wise quant support, these checkpoints cannot be loaded natively, and users miss out on the memory bandwidth savings that quantization provides during decode.

### Proposed Solution

1. **Block-wise quantized matmul kernel for `QuantizedLinear`** — support per-channel, block-channel/sub-channel, and 2D block scale formats, adapted from TPU Inference project.
2. **Block-wise quant support for `EPMoE`** — dynamic/static scale loading, dynamic/static quantization, and scale normalization to GMM kernel layout.
3. **Scale expansion at init/load time** — move `jnp.repeat` scale expansion from runtime to weight loading, avoiding per-step overhead in the forward pass.
4. **Quantization config YAML** — support configs like `int8_block_128_dynamic.yaml` and `int8_moe_block_128_linear_channel_dynamic.yaml` for flexible per-layer quantization control.

### Benchmark Results (TPU v6e-4)

**Qwen3-8B** (`tp=4`): INT8 block 128 decode throughput on par with BF16, ~27% prefill overhead from blockwise kernel.

**Qwen3-30B-A3B** (`tp=4, ep=4`): INT8 mixed (Linear per-channel + MoE block) decode throughput slightly better than BF16 at larger batch sizes, ~9% prefill overhead. Accuracy within noise of BF16 on GSM8K and GPQA Diamond.

### Related resources

- PR: #862

### PR DESCRIPTION
**Pull request by cjx0709**
## Motivation

Add block-wise quantization support from TPU inference, with coverage for both dense Linear layers and EPMoE.

Closes #879

## Modifications

- Add block-wise quant support for 'QuantizedLinear', basing on adapted quanted linear kernel code from TPU Inference project includeing:
  - per-channel
  - block-channel/ sub-channel
  - 2D block scale formats
- Add block-wise quant support for 'EPMoE', includeing:
  - dynamic/static scale loading
  - dynamic/static quantization
  - scale normalization to the GMM kernel layout 
- Add block-wise supported quantized matmul kernel for TPU inference
  - Integrate the TPU blockwise quantized matmul kernel with tuned fallback logic and safer validation checks.
- Move block-quant scale expansion (`jnp.repeat`) from runtime to init/load time for better inference performance

## Accuracy Tests
This PR will not effects for existing non-quantized model or channel-wise kernel. And add test List into CI about:
- Unit tests:
  - `python/sgl_jax/test/kernels/quantized_linear_test.py`
  - `python/sgl_jax/test/kernels/moe_block_quant_test.py`
  - `python/sgl_jax/test/test_linear_tp.py`
- TPU E2E tests:
  - `test/srt/test_moe_block_quant_e2e.py`
  - `test/srt/quantization/test_w8_block_dynamic_quantization.py`
  - `test/srt/quantization/test_w8_moe_block_linear_channel_quantization.py`

### Qwen3-30B-A3B (`tp=4, ep=4`)

| Eval Set | Variant | Score | Delta vs BF16 |
|---|---:|---:|---:|
| GSM8K | BF16 | 0.939 | - |
| GSM8K | INT8 mixed (`Linear per-channel + MoE block`) | 0.941 | +0.002 |
| GPQA Diamond | BF16 | 0.152 | - |
| GPQA Diamond | INT8 mixed (`Linear per-channel + MoE block`) | 0.141 | -0.011 |

### Qwen3-8B (`tp=4`)

| Eval Set | Variant | Score | Delta vs BF16 |
|---|---:|---:|---:|
| GSM8K | BF16 | 0.861 | - |
| GSM8K | INT8 block dynamic | 0.897 | +0.036 |
| GPQA Diamond | BF16 | 0.066 | - |
| GPQA Diamond | INT8 block dynamic | 0.121 | +0.055 |

## Benchmarking and Profiling

**Hardware:** TPU v6e-4
**Benchmark tool:** `bench_one_batch_server.py` (single-batch offline, input_len=4096, output_len=1024)
**Server config:** precompilation enabled

### Qwen3-8B BF16 (`tp=4`)

| batch size | latency (s) | input throughput (tok/s) | output throughput (tok/s) | ITL (ms) |
|--:|--:|--:|--:|--:|
| 1 | 5.91 | 48172.07 | 175.83 | 5.69 |
| 4 | 6.48 | 56459.25 | 661.91 | 6.04 |
| 8 | 7.10 | 57645.86 | 1254.03 | 6.38 |
| 16 | 8.77 | 58387.38 | 2142.96 | 7.47 |
| 32 | 12.28 | 58743.15 | 3261.23 | 9.81 |

### Qwen3-8B INT8 block dynamic (`tp=4`)

Config: `int8_block_128_dynamic.yaml`

| batch size | latency (s) | input throughput (tok/s) | output throughput (tok/s) | ITL (ms) |
|--:|--:|--:|--:|--:|
| 1 | 6.00 | 36538.96 | 173.85 | 5.75 |
| 4 | 6.29 | 41405.21 | 694.67 | 5.76 |
| 8 | 7.18 | 42115.50 | 1279.84 | 6.25 |
| 16 | 9.15 | 42494.77 | 2154.47 | 7.43 |
| 32 | 13.06 | 42679.25 | 3279.24 | 9.76 |

### Qwen3-30B-A3B BF16 (`tp=4, ep=4`)

| batch size | latency (s) | input throughput (tok/s) | output throughput (tok/s) | ITL (ms) |
|--:|--:|--:|--:|--:|
| 1 | 4.87 | 23956.37 | 217.74 | 4.59 |
| 4 | 6.10 | 25592.71 | 750.14 | 5.33 |
| 8 | 9.09 | 25871.30 | 1046.73 | 7.64 |
| 16 | 13.54 | 25951.58 | 1487.14 | 10.76 |
| 32 | 22.60 | 26033.21 | 1865.62 | 17.15 |

### Qwen3-30B-A3B INT8 mixed (`tp=4, ep=4`)

Config: `int8_moe_block_128_linear_channel_dynamic.yaml` (Linear per-channel + MoE block)

| batch size | latency (s) | input throughput (tok/s) | output throughput (tok/s) | ITL (ms) |
|--:|--:|--:|--:|--:|
| 1 | 5.56 | 21717.74 | 190.75 | 5.24 |
| 4 | 5.88 | 23351.42 | 790.99 | 5.06 |
| 8 | 8.07 | 23682.42 | 1224.88 | 6.53 |
| 16 | 12.87 | 23867.11 | 1617.74 | 9.89 |
| 32 | 22.91 | 23780.35 | 1883.15 | 16.99 |

### Key Observations

- **Qwen3-8B**: INT8 block 128 prefill throughput ~27% lower than BF16 (blockwise kernel overhead), but decode throughput is on par or slightly better due to reduced memory bandwidth from INT8 weights. ITL is nearly identical.
- **Qwen3-30B-A3B (MoE)**: INT8 mixed (Linear per-channel + MoE block) prefill throughput ~9% lower than BF16, decode throughput slightly higher at larger batch sizes. ITL comparable.
- **Accuracy**: Both quantized variants maintain accuracy within noise of BF16 baselines on GSM8K and GPQA Diamond.

## Checklist

- [x] Please use English, otherwise it will be closed.
- [x] The purpose of the PR, or link existing issues this PR will resolve.
- [x] The test plan, such as providing test command.
- [ ] (Optional) The necessary documentation update.

### CODE REVIEW

File: python/sgl_jax/srt/layers/linear.py [hunk 1]
```diff
@@ -204,58 +272,71 @@ def __call__(self, x: jax.Array) -> tuple[jax.Array, jax.Array | None]:
...
+        # After transpose from LinearBase, its sharding is
+        # P(kernel_axes[1], kernel_axes[0]).
+        #
+        # Input x sharding: P(None, input_axis)
         # Weight w_q sharding: P(output_axis, input_axis)
-        # Weight scale sharding: P(output_axis) - per output channel
+        # Weight scale sharding:
+        #   per-channel: P(output_axis)
+        #   block-wise:  P(output_axis, input_axis)
         # Output sharding: P(None, output_axis)
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/layers/linear.py]: Remove unnecessary comments.

File: python/sgl_jax/srt/kernels/quantized_matmul/kernel.py [hunk 2]
```diff
@@ -31,41 +97,124 @@ def xla_quantized_matmul_local(
...
+            out = lax.dot_general(
+                lhs,
+                w_dequant,
+                dimension_numbers=(((x.ndim - 1,), (1,)), ((), ())),
+                preferred_element_type=compute_dtype,
+            )
+            if quantize_activation:
+                out = out * x_scale.astype(compute_dtype)
+
     else:
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/kernels/quantized_matmul/kernel.py:100]: I don't think this fallback implementation should be used — it's too verbose. The TPU-inference check only needs:
`enable_quantized_matmul_kernel = len(w_s.shape) == 3`

File: python/sgl_jax/srt/kernels/quantized_matmul/3rd_quantized_matmul/__init__.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# Copyright 2025 Google LLC
+#
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/kernels/quantized_matmul/3rd_quantized_matmul/__init__.py]: I don't think the name needs to indicate it's a third-party implementation — it's not currently a Python third-party library.

File: python/sgl_jax/srt/kernels/quantized_matmul/kernel.py (hunk 2 above)
Comment (cjx0709) [reply python/sgl_jax/srt/kernels/quantized_matmul/kernel.py:100]: I use verbose if-else logic because I would to support block-wise quant for non-tpu envrionment, my origin logic: 
1. w_scale == 3 and no tpu env:
	using jax block-wise quant;
2. w_scale == 3 and tpu env:
	using 3rd block-wise palles kernel;
3. w_scale != 3:
	using jax channel-wise kernel.
But it seems too complex, I will follow tpu-inference, delete case 1, only support case 2 and case 3 previous.

Comment (Bob-Chen222) [reply python/sgl_jax/srt/kernels/quantized_matmul/kernel.py:100]: Just make sure I'm following, what case to handle w_scale==2 and tpu env if deleted second case?

Comment (cjx0709) [reply python/sgl_jax/srt/kernels/quantized_matmul/kernel.py:100]: sorry for my error write，I mean I will delete case 1. @Bob-Chen222

File: python/sgl_jax/srt/layers/linear.py (hunk 1 above)
Comment (cjx0709) [reply python/sgl_jax/srt/layers/linear.py]: done

File: python/sgl_jax/srt/kernels/quantized_matmul/3rd_quantized_matmul/__init__.py (hunk 3 above)
Comment (cjx0709) [reply python/sgl_jax/srt/kernels/quantized_matmul/3rd_quantized_matmul/__init__.py]: adjusted

File: python/sgl_jax/srt/kernels/quantized_matmul/kernel.py [hunk 4]
```diff
@@ -25,47 +32,103 @@ def xla_quantized_matmul_local(
...
+        in_blocks = w_scale.shape[0]
+        block_size_in = in_dim // in_blocks
+
+        if weight_block_size is not None:
+            block_size_out = int(weight_block_size[0])
+        else:
+            block_size_out = block_size_in
+
+        blockwise_kernel = get_blockwise_kernel()
+        if jax.default_backend() != "tpu":
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/kernels/quantized_matmul/kernel.py]: Break on initialization

File: python/sgl_jax/srt/layers/moe.py
Comment (Bob-Chen222) [inline python/sgl_jax/srt/layers/moe.py:1]: Maybe better to separate different components into different files

File: python/sgl_jax/srt/kernels/quantized_matmul/kernel.py (hunk 4 above)
Comment (cjx0709) [reply python/sgl_jax/srt/kernels/quantized_matmul/kernel.py]: done

File: python/sgl_jax/srt/layers/moe.py
Comment (cjx0709) [reply python/sgl_jax/srt/layers/moe.py:1]: already split moe.py to 3 file which all less than 1000 lines.

File: python/sgl_jax/srt/kernels/quantized_matmul/kernel.py (hunk 4 above)
Comment (cjx0709) [reply python/sgl_jax/srt/kernels/quantized_matmul/kernel.py]: I tried，but will meet：
image
it happends because   jax.default_backend() triggers JAX TPU backend initialization and acquires an exclusive TPU lock. This call was moved from kernel.py (scheduler-only code) into quantization_config.py:from_yaml( (shared code called by both processes) in commit 270a39fd:

 int quantization_config.py, line 167
  if weight_block_size is not None and jax.default_backend() != "tpu":
      raise RuntimeError(...)

  The server uses a multi-process architecture: the tokenizer process (parent) spawns a scheduler subprocess (child). Only the scheduler should hold the TPU lock. However, the tokenizer process also calls from_yaml() during initialization:
 engine.py:_launch_subprocesses()
    → TokenizerManager(server_args)                   # parent process
      → ModelConfig.from_server_args(server_args)
        → QuantizationConfig.from_path(config_path)
          → from_yaml(yaml_path)
            → jax.default_backend()                    # ← initializes TPU in parent, grabs lock

  When the scheduler subprocess subsequently calls jax.devices(), it finds the TPU is already held by the parent process (pid 6871) and crashes.

May can't use check when initializing.

Comment (JamesBrianD) [reply python/sgl_jax/srt/kernels/quantized_matmul/kernel.py]: I think we should move the validation to model init. That usually runs on the scheduler process anyway, so we can fail early if the block-wise config or backend is incompatible.

Comment (cjx0709) [reply python/sgl_jax/srt/kernels/quantized_matmul/kernel.py]: Now it will raise error when load_model if block-wise config and backend are incompatible.
image

Comment (JamesBrianD) [reply python/sgl_jax/srt/kernels/quantized_matmul/kernel.py]: LGTM

**Review by JamesBrianD — APPROVED** [review]
